nucliadb_node/src/shards/shard_writer.rs

```diff
@@ -236,135 +240,196 @@ impl ShardWriter {
+        Ok(())
+    }
+
+    pub fn remove_vectors_index(&self, name: String) -> NodeResult<()> {
+        let mut indexes = ShardIndexes::load(&self.metadata.shard_path())?;
+        let path = indexes.remove_vectors_index(&name)?;
+        indexes.store()?;
+        write_rw_lock(&self.indexes).vectors_indexes.remove(&name);
+        if let Some(path) = path {
+            // Although there can be a reader with this index open, readers
+            // currently open all vectors index files so we rely on Linux not
+            // deleting the files until it closes the index. Readers then should
+            // be able to keep answering for that vectorset until closing it
+            std::fs::remove_dir_all(path)?;
+        }
+        Ok(())
+    }
+
+    pub fn list_vectors_indexes(&self) -> Vec<String> {
+        read_rw_lock(&self.indexes).vectors_indexes.keys().cloned().collect::<Vec<String>>()
+    }
+
     #[measure(actor = "shard", metric = "set_resource")]
     #[tracing::instrument(skip_all)]
     pub fn set_resource(&self, mut resource: Resource) -> NodeResult<()> {
         let span = tracing::Span::current();
 
         remove_invalid_labels(&mut resource);
 
-        let indexes: &mut InnerShardWriter = &mut write_rw_lock(&self.indexes);
+        let indexes: &mut ShardWriterIndexes = &mut write_rw_lock(&self.indexes);
 
-        let text_task = || {
-            debug!("Field service starts set_resource");
-            let result = indexes.text_writer.set_resource(&resource);
-            debug!("Field service ends set_resource");
-            result
+        let mut text_task = || {
+            run_with_telemetry(info_span!(parent: &span, "text set_resource"), || {
+                debug!("Field service starts set_resource");
+                let result = indexes.texts_index.set_resource(&resource);
+                debug!("Field service ends set_resource");
+                result
+            })
         };
 
-        let paragraph_task = || {
-            debug!("Paragraph service starts set_resource");
-            let result = indexes.paragraph_writer.set_resource(&resource);
-            debug!("Paragraph service ends set_resource");
-            result
+        let mut paragraph_task = || {
+            run_with_telemetry(info_span!(parent: &span, "paragraph set_resource"), || {
+                debug!("Paragraph service starts set_resource");
+                let result = indexes.paragraphs_index.set_resource(&resource);
+                debug!("Paragraph service ends set_resource");
+                result
+            })
         };
 
-        let vector_task = || {
-            debug!("Vector service starts set_resource");
-            let result = indexes.vector_writer.set_resource(&resource);
-            debug!("Vector service ends set_resource");
-            result
-        };
+        let mut vector_tasks = vec![];
+        for (_, vector_writer) in indexes.vectors_indexes.iter_mut() {
```

Comment:
Won't this write the same vector to all vectorsets?

I guess it's still a placeholder until we change the SetResource message?

Reply:
Yes, maybe that's a good opportunity to clean protobuffers and pass a custom struct instead of the whole `Resource` to `nucliadb_vectors`